packages/core/src/protocol-interfaces.ts

```diff
@@ -55,6 +55,7 @@ export interface ProtocolClientFactory {
 export interface ProtocolServer {
   readonly scheme: string;
   expose(thing: ExposedThing, tdTemplate?: WoT.ThingDescription): Promise<void>;
+  destroy(thingId: string): Promise<boolean>; // reports success
```

Comment:
I am wondering if it makes more sense to throw if the thing was not destroyed instead of returning false. Consider also if we just return false we can't handle the causes from the code, but we probably need to check the logs to understand why.

Reply:
Mhh, the reason why I came up with a boolean is that some bindings might not even support destroying (for example some bindings like websocket/oracle do not have an implementation yet and can indicate that destroy did not work)

Moreover, the servient that calls all bindings to destroy() a given thing can decide what do do.. throwing to me means a failure happened while `false` to me means it did not work...
Does that make sense?

Reply:
It is reasonable, we just need to think about it thoughtfully. So from the servient point of (the call of each `destroy` function), I can expect two unwanted outcomes:
- The function returned `false`
- The function throws an error

Can we define two different behaviors for those conditions? From my point of view, I would just assume that the service that I wanted (removing the thing) was not performed. So it feels like an error state, isn't it? 

About the `not implemented` situation: does it makes sense to have bindings that do not ever implement the `destroy` functionality? Otherwise, returning `false` is just a //TODO: placeholder, right?

Reply:
> Can we define two different behaviors for those conditions? From my point of view, I would just assume that the service that I wanted (removing the thing) was not performed. So it feels like an error state, isn't it?

Let's assume the binding.destroy() failed (due to what every reason). Should the servient.destroy() remove the thing from the list or not? I am not sure here either ....

> About the `not implemented` situation: does it makes sense to have bindings that do not ever implement the `destroy` functionality? Otherwise, returning `false` is just a //TODO: placeholder, right?

is possible that a binding can expose a thing but not destroying it again. Not sure.

Reply:
> Let's assume the binding.destroy() failed (due to what every reason). Should the servient.destroy() remove the thing from the list or not? I am not sure here either ....

Hmm not a pleasant state, indeed. I have a feeling that removing it would be wrong... but still if we keep it who knows which is the actual state of the thing... it might be half-working like receiving requests only for one affordance type. In my opinion, we should check what `sockets` do when you close them. 

> is possible that binding can expose a thing but not destroying it again. Not sure.
So it creates it, but not destroys it... IMHO it would be better than if it creates resources for exposing it MUST also clean its own mess.

Reply:
I am not sure how to proceed here, mhh.
@egekorkan do you have any opinion?

Reply:
I am not sure in which cases the destroy would not work but having the boolean as return value creates some flexibility in case we want to communicate such cases

Reply:
@relu91 @egekorkan  do we agree to keep the interface and the code as is?

Reply:
I'd say yes for boolean interface above

Reply:
> I am not sure in which cases the destroy would not work but having the boolean as return value creates some flexibility in case we want to communicate such cases

Exactly because we are not sure which are the cases when a destroy function would return false (i.e., fail) I would prefer using void. When design functions I follow this rule:

> The rule is pretty simple: Use exceptions on exceptional, erroneous, unpredicted failures. Don't use exceptions when you expect something to happen or when something happens really often.

Are we _expecting_ that the destroy function would return a boolean value? Looking around the close functions return void and throw if something is wrong. I would follow this pattern, also considering the principle of the least surprise. 

[Java](https://docs.oracle.com/javase/7/docs/api/java/io/Closeable.html)
[Nodejs](https://nodejs.org/api/stream.html#stream_writable_destroy_error)

Reply:
I see your argument and agree with it now. Letting a failure be reported with a boolean=false would also "encourage" developers to simplify/abstract the information related to failure which would not be very helpful to detect the failure.

Reply:
It seems I have the opposite opinion 🙈

* I suggest to throw errors if some code part fails or an error occurs
* in this case a binding may choose not support removing/destroying a thing (hence returning false seems ok to me)

Currently we have bindings like fujitsu, oracle, ws that have not implemented an actual `destroy()`. At the moment the call succeeds and returns `false` to indicate no destroy/remove took place. IF we decide to go with throws I think we loose the information that there isn't an issue but it rather did not work...

Similarly, what should happen if a provided thingId was not found. `throw` an error or indicate `false`?

To me it is very similar to removing an entry from a list.
see https://docs.oracle.com/javase/8/docs/api/java/util/List.html#remove-java.lang.Object-
It may return boolean or throw in case of failure.

Reply:
Question: let's assume we go with void and `ExposedThing.destroy()` walks over all bindings and calls `destroy(thingId)`. What should happen it it fails for one binding?

Reply:
> in this case a binding may choose not support removing/destroying a thing (hence returning false seems ok to me)

Most implementations use the famous [NotImplementedExecption](https://softwareengineering.stackexchange.com/questions/323498/why-is-having-a-notimplementedexception-a-good-thing), which is a thing also in TS codebases. 

As I said above a binding implementation must do something in the destroy method, cause in principle an ExposedThing is occupying system resources. Even in ws situations, you might need to close all the open connections, right?

> Similarly, what should happen if a provided thingId was not found. throw an error or indicate false?
>
>Similarly, what should happen if a provided thingId was not found. throw an error or indicate false?
>
>To me it is very similar to removing an entry from a list.
see https://docs.oracle.com/javase/8/docs/api/java/util/List.html#remove-java.lang.Object-
It may return boolean or throw in case of failure.

Now I see, but I would explicitly use the return value only to indicate that a Web Thing was not found. We should write that in jsdoc and not use it for other errors (like not implemented above). Something like:
```js
/**
 * @param thingId: id of the thing to destroy 
 * @returns true if the thing was found and destroyed; false if the thing was not in the list
 * @throws if the binding couldn't destroy completely the thing with thingId.
**/
function destroy(thingId: string): Promise<boolean>;
```

I would still argue that from a consumer point of view the returned value is not so important. For example, if I want to remove an element from the list if it wasn't there I still obtain the effect that the *element is not in the list*. Some way here, if I try to destroy a thing that was not in the binding, well *nothing to be cleaned*. But I guess you got a point.

Reply:
Looks very promising to me!
I think we get close to agreement.

Some final questions:
* for the oracle-server I do no not really know what to do fro destroy(). Hence I would like to throw `NotImplementedException`. Is this really a error we should use, not really defined in https://nodejs.org/api/errors.html
* another question (the same as my [comment above](https://github.com/eclipse/thingweb.node-wot/pull/438#discussion_r671114750)). Let's assume ExposedThing.destroy() walks over all bindings and calls destroy(thingId). What should happen it it fails for one binding (e.g., the oracle binding)? The [current code](https://github.com/eclipse/thingweb.node-wot/blob/a39fbc643248863ee5c2d376e1550af7e83b8d32/packages/core/src/exposed-thing.ts#L109-L133) would reject since a error was thrown... or shall we look into the error type and ignore `NotImplementedException` ?

Reply:
I think that the main argument should be based on how the Scripting level code should look like. If the code can do with the knowledge of the boolean value, we can have it. Mostly, the code wants to simply destroy the Thing and move on and not handle true, false and errors independently no?

Reply:
> I think that the main argument should be based on how the Scripting level code should look like. If the code can do with the knowledge of the boolean value, we can have it. Mostly, the code wants to simply destroy the Thing and move on and not handle true, false and errors independently no?

Yes, correct, but here we are discussing on the internal level Core <-> Bindings. `ExposeThing.destroy()` returns `void`. 

> for the oracle-server I do no not really know what to do fro destroy(). Hence I would like to throw NotImplementedException. Is this really a error we should use, not really defined in https://nodejs.org/api/errors.html

I'd go for something like this:
```ts
function _notImplemented() : never {
     throw new Error("Not Implemented")
}
```

> another question (the same as my comment above). Let's assume ExposedThing.destroy() walks over all bindings and calls destroy(thingId). What should happen it it fails for one binding (e.g., the oracle binding)? The current code would reject since a error was thrown... or shall we look into the error type and ignore

Be transparent with the application code if something fails give more control and leave room for customization. I'd throw if _at least_ one binding fails. Like go through every biding, collect errors, if `errors.length >0` -> throw. What do you think?